bottlecap/src/appsec/processor/mod.rs

```diff
@@ -247,11 +260,375 @@ pub trait InvocationPayload {
+            Self::APIGatewayRestEvent(t) => t.query_parameters.clone(),
+            Self::APIGatewayWebSocketEvent(t) => t.query_parameters.clone(),
+            Self::ALBEvent(t) => {
+                if t.multi_value_headers.is_empty() {
+                    t.headers
+                        .iter()
+                        .map(|(k, v)| (k.clone(), vec![v.clone()]))
+                        .collect()
+                } else {
+                    t.multi_value_headers.clone()
+                }
+            }
+            Self::LambdaFunctionUrlEvent(t) => t
+                .query_string_parameters
+                .iter()
+                .map(|(k, v)| (k.clone(), vec![v.clone()]))
+                .collect(),
+            // Events that are not relevant to App & API Protection
+            Self::MSKEvent(_)
+            | Self::SqsRecord(_)
+            | Self::SnsRecord(_)
+            | Self::DynamoDbRecord(_)
+            | Self::S3Record(_)
+            | Self::EventBridgeEvent(_)
+            | Self::KinesisRecord(_)
+            | Self::StepFunctionEvent(_)
+            | Self::Unknown => HashMap::default(),
+        }
+    }
+    fn path_params(&self) -> HashMap<String, String> {
+        match self {
+            Self::APIGatewayHttpEvent(t) => t.path_parameters.clone(),
+            Self::APIGatewayRestEvent(t) => t.path_parameters.clone(),
+            Self::APIGatewayWebSocketEvent(t) => t.path_parameters.clone(),
+            #[allow(clippy::match_same_arms)]
+            Self::ALBEvent(_) => HashMap::default(),
+            #[allow(clippy::match_same_arms)]
+            Self::LambdaFunctionUrlEvent(_) => HashMap::default(),
+            // Events that are not relevant to App & API Protection
+            Self::MSKEvent(_)
+            | Self::SqsRecord(_)
+            | Self::SnsRecord(_)
+            | Self::DynamoDbRecord(_)
+            | Self::S3Record(_)
+            | Self::EventBridgeEvent(_)
+            | Self::KinesisRecord(_)
+            | Self::StepFunctionEvent(_)
+            | Self::Unknown => HashMap::default(),
+        }
+    }
+    fn request_body<'a>(&'a self) -> Option<Box<dyn Read + 'a>> {
+        match self {
+            Self::APIGatewayHttpEvent(t) => t.body.reader().ok().flatten(),
+            Self::APIGatewayRestEvent(t) => t.body.reader().ok().flatten(),
+            Self::APIGatewayWebSocketEvent(t) => t.body.reader().ok().flatten(),
+            Self::ALBEvent(t) => t.body.reader().ok().flatten(),
+            Self::LambdaFunctionUrlEvent(t) => t.body.reader().ok().flatten(),
+            // Events that are not relevant to App & API Protection
+            Self::MSKEvent(_)
+            | Self::SqsRecord(_)
+            | Self::SnsRecord(_)
+            | Self::DynamoDbRecord(_)
+            | Self::S3Record(_)
+            | Self::EventBridgeEvent(_)
+            | Self::KinesisRecord(_)
+            | Self::StepFunctionEvent(_)
+            | Self::Unknown => None,
+        }
+    }
+}
+
```

Comment:
Wondering if some of these methods should be struct methods for the events/triggers that we care about, although I guess the decision for this was that some of them won't actually have the method?

Reply:
Yeah I _could_ have implemented this trait on the trigger structs themselves; and just have had the `IdentifiedTriggers` have a `fn into_trigger(self) -> Option<Box<dyn InvocationPayload>>` instead...

It kind of made sense for me to have these implementations under the `appsec` module as they're sensitive to how appsec thinks of these concepts... but I'm not feeling all too strongly about this aspect of it.